spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/handler/predicate/OpenGrayRoutePredicateFactory.java

```diff
@@ -0,0 +1,195 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gateway.handler.predicate;
+
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+import org.apache.commons.codec.digest.MurmurHash3;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.http.HttpCookie;
+import org.springframework.util.ObjectUtils;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.server.ServerWebExchange;
+
+/**
+ * other than WeightRoutePredicateFactory route traffic random,
+ * this predicate route traffic base on specific parameter
+ */
+
+public class OpenGrayRoutePredicateFactory extends AbstractRoutePredicateFactory<OpenGrayRoutePredicateFactory.Config> {
+	private static final Log log = LogFactory.getLog(OpenGrayRoutePredicateFactory.class);
+    private final String REQUEST_QUERY_PATTERN = "request.query.";
+    private final String REQUEST_HEADER_PATTERN = "request.header.";
+    private final String REQUEST_COOKIE_PATTERN = "request.cookie.";
+    private final String REQUEST_PATH_PATTERN = "request.path";
+
+    /**
+     * available range is [0, 1000)
+     */
+    public static final int MAX_WEIGHT_RANGE = 1000;
+
+    public static final String FILED_KEY = "fieldPattern";
+    /**
+     */
+    public static final String START_KEY = "start";
+    public static final String END_KEY = "end";
+
+    public static final String MATCH_EMPTY_FIELD_KEY = "matchEmptyField";
+
+
+    public OpenGrayRoutePredicateFactory() {
+        super(Config.class);
+    }
+
+    @Override
+    public List<String> shortcutFieldOrder() {
+        return Arrays.asList(FILED_KEY, START_KEY, END_KEY, MATCH_EMPTY_FIELD_KEY);
+    }
+
+    @Override
+    public Predicate<ServerWebExchange> apply(Config config) {
+        return new GatewayPredicate() {
+            @Override
+            public boolean test(ServerWebExchange exchange) {
+                // get gray value
+                String patternValue = patternParamResolve(exchange, config.getFieldPattern());
+                if (ObjectUtils.isEmpty(patternValue)) {
+                    return config.isMatchEmptyField();
+                }
+
+
+                int weightStart = Math.max(0, config.getStart());
+				int weightEnd = Math.min(MAX_WEIGHT_RANGE, config.getEnd());
+
+                if (weightEnd < weightStart) {
+                    // invalid match range
+                    return false;
+                }
+
+
+				int hash = Math.abs(MurmurHash3.hash32x86(patternValue.getBytes(StandardCharsets.UTF_8))% MAX_WEIGHT_RANGE) ;
```

Comment:
Let's create an interface that extends `Function<byte[], int>` and require a bean of that type for this to work. That way the use of apache commons doesn't leak here.